pkg/reconciler/parallel/parallel.go

```diff
@@ -254,3 +252,86 @@ func (r *Reconciler) trackAndFetchChannel(ctx context.Context, p *v1.Parallel, r
 	}
 	return obj, err
 }
+
+func (r *Reconciler) removeUnwantedChannels(ctx context.Context, channelResourceInterface dynamic.ResourceInterface, p *v1.Parallel, wanted []*duckv1.Channelable) error {
+	channelObjRef := corev1.ObjectReference{
+		Kind:       p.Spec.ChannelTemplate.Kind,
+		APIVersion: p.Spec.ChannelTemplate.APIVersion,
+	}
+
+	l, err := r.channelableTracker.ListerFor(channelObjRef)
+	if err != nil {
+		logging.FromContext(ctx).Errorw("Error getting lister for Channel", zap.Any("channelRef", channelObjRef), zap.Error(err))
+		return err
+	}
+
+	exists, err := l.ByNamespace(p.GetNamespace()).List(labels.Everything())
+	if err != nil {
+		logging.FromContext(ctx).Errorw("Error listing Channels", zap.Any("namespace", p.Namespace), zap.Any("channelRef", channelObjRef), zap.Error(err))
+		return err
+	}
+
+	for _, c := range exists {
+		ch, err := kmeta.DeletionHandlingAccessor(c)
+		if err != nil {
+			logging.FromContext(ctx).Errorw("Failed to get channel", zap.Any("channel", c), zap.Error(err))
+			return err
+		}
+
+		if !ch.GetDeletionTimestamp().IsZero() ||
+			!metav1.IsControlledBy(ch, p) {
+			continue
+		}
+
+		used := false
+		for _, cw := range wanted {
+			if cw.Name == ch.GetName() {
+				used = true
+				break
+			}
+		}
```

Comment:
A general observation about this method:

It might be more efficient and clearer to just initialize two sets of Channel names:
- one with all the "wanted" ones
- one with the ones we "own"

Then delete all channels returned by [`owned.Difference(wanted)`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets?utm_source=godoc#String.Difference).